core/src/server/rpc_module.rs

```diff
@@ -869,68 +971,72 @@ impl SubscriptionSink {
 	}
 
-	fn build_message<T: Serialize>(&self, result: &T) -> Result<String, Error> {
+	fn is_active_subscription(&self) -> bool {
+		Arc::strong_count(&self.active_sub) > 1
+	}
+
+	fn build_message<T: Serialize>(&self, result: &T) -> Result<String, serde_json::Error> {
 		serde_json::to_string(&SubscriptionResponse::new(
 			self.method.into(),
 			SubscriptionPayload { subscription: self.uniq_sub.sub_id.clone(), result },
 		))
 		.map_err(Into::into)
 	}
 
-	fn inner_send(&mut self, msg: String) -> Result<(), Error> {
-		let res = match self.is_connected.as_ref() {
-			Some(conn) if !conn.is_canceled() => {
-				// unbounded send only fails if the receiver has been dropped.
-				self.inner.send_raw(msg).map_err(|_| Some(SubscriptionClosedReason::ConnectionReset))
-			}
-			Some(_) => Err(Some(SubscriptionClosedReason::Unsubscribed)),
-			// NOTE(niklasad1): this should be unreachable, after the first error is detected the subscription is closed.
-			None => Err(None),
-		};
-
-		// The subscription was already closed by the client
-		// Close down the subscription but don't send a message to the client.
-		if res.is_err() {
-			self.inner_close(None);
-		}
-
-		res.map_err(|e| {
-			let err = e.unwrap_or_else(|| SubscriptionClosedReason::Server("Close reason unknown".to_string()));
-			Error::SubscriptionClosed(err.into())
-		})
-	}
-
-	/// Close the subscription sink with a customized error message.
-	pub fn close_with_custom_message(&mut self, msg: &str) {
-		let close_reason = SubscriptionClosedReason::Server(msg.to_string()).into();
-		self.inner_close(Some(&close_reason));
+	fn build_error_message<T: Serialize>(&self, error: &T) -> Result<String, serde_json::Error> {
+		serde_json::to_string(&SubscriptionError::new(
+			self.method.into(),
+			SubscriptionPayloadError { subscription: self.uniq_sub.sub_id.clone(), error },
+		))
+		.map_err(Into::into)
 	}
 
-	/// Close the subscription sink with the provided [`SubscriptionClosed`].
-	pub fn close(&mut self, close_reason: &SubscriptionClosed) {
-		self.inner_close(Some(close_reason));
+	fn inner_send(&mut self, msg: String) -> bool {
+		if self.is_active_subscription() {
+			self.inner.send_raw(msg).is_ok()
+		} else {
+			false
+		}
 	}
 
-	fn inner_close(&mut self, maybe_close: Option<&SubscriptionClosed>) {
-		self.is_connected.take();
-		if let Some((sink, _)) = self.subscribers.lock().remove(&self.uniq_sub) {
-			tracing::debug!("Closing subscription: {:?} reason: {:?}", self.uniq_sub.sub_id, maybe_close);
-
-			let closed_by_server =
-				maybe_close.map_or(false, |c| matches!(c.close_reason(), SubscriptionClosedReason::Server(_)));
+	/// Send an error notification on the subscription with a special field `error`.
+	///
+	/// This may be used for indicating to connected clients that a given subscription has been canceled
+	/// as well so not necessarily an error but this is used for compat reason with existing JavaScript libraries.
```

Comment:
```suggestion
	/// Close the subscription, sending a notification with a special `error` field containing the provided error.
	///
	/// This can be used to signal an actual error, or just to signal that the subscription has been closed,
	/// depending on your preference. 
```